Fixes #21. Plot source estimates using ipyvolume.

OlehKSS opened the pull request:
Function for plotting source estimates using ipyvolume. In this version it does everything previously created functions for visualizing raw data do.

So far, most of the features of original function `plot_source_estimates` have been implemented.

All implemented:
- [x] time_label,
- [x] time_viewer,
- [x] initial_time,
- [x] time_unit.
- [x] cortex, it not possible to provide any options different from the default one
- [x] foreground: used for text color (so far), *implemented*
- [x] views as a list added
- [x] colorbar, *done with bqplot*
- [x] title of the figure, *added*.
- [x]  the figure parameter was *added*, so far it will raise `NotImplementedError`.

Not added on purpose:
- [x] spacing(can be applied to the matplotlib-based solution only), omitted.

larsoner commented:
> I have a question regarding SourceEstimates.data property, the shape of that numpy array is around 30000 by 25, now I am doing stc.crop(0.09, 0.09) and select 0th column stc.data[:, 0]. Is it correct approach?

You shouldn't need to do any `.crop` operation. Based on the time index the user has selected, you display the correct time index with something like `stc.data[:, time_idx]`.

> Another question regarding the SourceEstimate data type, should the number of vertices of the mesh be equal to the number of data points of stc object?

In general no, there should be an order of magnitude fewer in `stc` than in the source space mesh. This is what the `smoothing_steps` parameter does in PySurfer. We have code to compute the upsampling matrix in MNE using http://www.martinos.org/mne/stable/generated/mne.compute_morph_matrix.html with something like:
```
mat = compute_morph_matrix(subject, subject, stc.vertices, [np.arange(len(lh_surf['rr'])), np.arange(len(rh_surf['rr']))], smoothing_steps, subjects_dir)
```
then to upsample you do `mat.dot(stc.data[:, time_idx])` because `mat` is a SciPy sparse matrix.

For the other params, play with PySurfer until you are comfortable. But here is a short version:

* time_label : text in the lower right of the figure that is updated to show the time
* time_viewer : in PySurfer this is a new window, for you it should be an IPython Widget slider
* initial_time : time in `time_unit` (s or ms) to show (can figure out the index using `stc.times`
* time_unit : `'s'` or `'ms'`, how to convert to times
* cortex : the MNE docstring is outdated, see http://pysurfer.github.io/generated/surfer.Brain.html#surfer.Brain and adapt the code from it
* foreground : color used for things in the foreground, usually just used for text objects (if you set `background='white'` then `foreground='auto'` will be `foreground='k'`, and if `background='k'` then it will use `foreground='w'`, etc. -- the logic for this lives in PySurfer, you can take it
* spacing : only used for `matplotlib` backend, don't bother with this
* colorbar : for now just omit it, please open an issue to add it, it will probably take a bit of work / potential collaboration with IPyvolume folks
* title of the figure : we don't use this in `stc.plot`
* figure : omit this, I think we always want to create a new one for now

Also, please add an issue for how we want to deal with multiple views and `hemi`. In PySurfer you can do `hemi='lh' | 'rh' | 'both' | 'split'`, as well as `views='lat'` or `views=['lat', 'med']`. Play with it until you get a sense of this. Hopefully IPyvolume gives you some way of having multiple sub-panels/sub-scenes in a single rendering context (i.e., a 2x2 grid where each panel in the grid has its own objects and camera controls). If it already has these mechanisms the translation should be straightforward. If it does not, then please open an IPyvolume issue discussing this use case so we can discuss potential solutions.

OlehKSS commented:
I have added two issues according to the comments above: #27, #26.
By looking at ipyvolume examples, I am not sure whether there will be a need to use a foreground color for ipyvolume backend. Well, it doesn't seem to be of the biggest priority so far. 

As far as I understand from `stc.data[:, time_idx]` , `time_idx` is an epoch, right?
So, I will need to add selection of proper `time_idx` and timing functions, options for the cortex (seems like a lot of them).

larsoner commented:
cortex is a very low priority

time_idx is the index corresponding to a particular time point in the
sources x time array

In MNE "epoch" refers to a channels x time array (basically a trial),
typically one of set of epochs in an array of shape n_epochs x channels x
time

>

OlehKSS commented:
Ok, I have crossed cortex out of my list, and I have added an issue for it #28.

OlehKSS commented:
I have added `initial_time` parameter support for `plot_source_estimates` function. I have also support for `time_viewer=True`, so this pull request will also fix #22.  Am I missing anything else? Can we do a code review in case everything looks good?

*Example of cortex graph with time scale enalbled*

![image](https://user-images.githubusercontent.com/16047980/41916353-2ab508f2-7958-11e8-8ca5-af1ca718c4de.png)

agramfort commented:
nice !

how do I test this? is it fast to interact?

OlehKSS commented:
@agramfort Use this [notebook](https://github.com/OlehKSS/mne-gsoc2018-3d/blob/olehkss/21/examples/plot_stc.ipynb) for testing.

I am using ipyvolume package installation from source code. Moreover, I am awaiting [my pull request](https://github.com/maartenbreddels/ipyvolume/pull/145) to ipyvolume with WebGL configuration changes to be merged. So, I suggest doing installation from that branch, since otherwise you will have a plot slightly different from what I have.

OlehKSS commented:
And it quite fast to interact

OlehKSS commented:
I have changed slider label to show correct time instance 

![image](https://user-images.githubusercontent.com/16047980/41922410-527e47ec-7965-11e8-9e6e-739e13173ef1.png)

larsoner commented:
Great. Eventually we want this `time_label` to be used the same way that it is in `stc.plot`, which IIRC means it will respect the `time_unit` parameter, too

OlehKSS commented:
I have fixed time label and added `time_label` parameter, so now this behavior corresponds to the behavior of `plot_source_estimate` function

![image](https://user-images.githubusercontent.com/16047980/41931345-1ae817a4-797e-11e8-966a-9a363aa12e90.png)

larsoner commented:
Works great, and is smooth on my workstation.

@OlehKSS maybe the next thing to do is to complete the set of arguments for the `plot_source_estimates` API. For arguments that are not supported with this backend yet, `raise NotImplementedError` (e.g., if `cortex is not True` or `hemi == 'split'`). You'll probably need to add support for `hemi='lh'` and `hemi='rh'` (but not yet `hemi='split'`), as well as the set of `views` that PySurfer supports, which should be easy enough.

Once we have that, perhaps it is time to actually move the code over to `mne-python` as an experimental feature?

agramfort commented:
I see this on my system:

<img width="522" alt="screen shot 2018-06-26 at 22 36 24" src="https://user-images.githubusercontent.com/161052/41937749-75aea4ba-7991-11e8-8804-a806c25a249f.png">

any idea what's going on? I am using PR/145 from ipyvolume as you suggested.

larsoner commented:
Do the main `ipyvolume` examples work? I'm guessing not.

What browser? IIRC Chrome tends to work best.

agramfort commented:
indeed they were not. I managed to fix it in another conda env. Really nice
!

I would be great to expose sliders to set the colorbar fmin, fmid and fmax.
And I also agree with @larsoner that using bqplot for the colorbar would
really
bring us close to a fully functional viewer in the notebook.

great job.

OlehKSS commented:
I have added `NotImplementedError` for `cortex`, `hemi="split"`, `foreground`, `colorbar=True`, and for unsupported values (`np.array`) of `colormap`. I have also added support for `time_label` of callable type. `hemi="lh" | "rh"` was implemented before, as well as `views`. I experience problems with  changing views using ipyvolume, it seems like a bug to me, because ipyvolume graph does not change its view at all, no matter which arguments are provided. I have tested on my old code which was working before,  I do not obtain the same results now.

OlehKSS commented:
As for adding a new backend to `MNE:plot_source_estimates`, I think it can be added in the same way as `_plot_mpl_stc` function. E. g. I can add a `_plot_ipv_stc` function, which will be similar to what I have now implemented in this pull request. Should I make a pull request to `MNE` repository afterwards?

larsoner commented:
Let's see if we can get the `bqplot` colorbar interaction working, then yeah a PR to `mne-python` adding a private function would work. Then future PRs would probably be made directly to `mne-python`

OlehKSS commented:
It seems like `bqplot` has no support for custom color maps and transparent colors. I was trying to create a color bar using heatmap and found this problem. It is possible to provide a list of several colors (seems like no more than three entries). There is no 'hot' color scheme available in `bqplot`. Well, it seems like `bqplot` support only color maps available in `d3.js` and several more, [list of color maps](https://github.com/bloomberg/bqplot/blob/master/js/src/colorbrewer.js). I asked `bqplot` guys for clarification. Nevertheless, it seems like `matplotlib` suits more to what we need.

choldgraf commented:
One thing you could probably do is to just use the numpy array color outputs from the colormaps in matplotlib and build a bqplot colorscale out of it: http://bqplot.readthedocs.io/en/stable/_generate/bqplot.scales.ColorScale.html

OlehKSS commented:
`ColorScale` constructor of `bqplot` accepts only a list of strings of colors specified in hexidecimal format. Output values from colormaps in matplotlib can be converted to comply with that type of format. Nevertheless, I see several problems so far. First of all, `ColorScale` doesn't support transparency channel, what to do with transparent colors in our case? Second, it doesn't seem to support centered colormaps  and it takes only first three colors for creating an instance of `ColorScale` class.

OlehKSS commented:
I have been also trying to implement colormap with `matplotlib`, the problem I am facing now is that a plot done with `matplotlib` cannot be easily put together with existing ipywidgets. One way to overcome it is to create a png-image out of the plot and show it using `Ipywidgets:Image` widget. Another way to visualize data is using d3js in Jupyter and write code for visualization in `HTML/CSS/JavaScript`. I haven't tried this approach yet. Is there any other library for interactive data visualization similar to `bqplot` I can check?

agramfort commented:
what do ipyvolume / bqplot folks suggest?

larsoner commented:
Transparency isn't a problem, you can use the underlying cortex average color (usually gray) and do the "transparency" mixing yourself. This is what we do already in PySurfer.

As for bqplot etc. I thought it was possible to have matplotlib interactive plots with widgets? If so then a colorbar should work.

As for Gitter, many times it is not actively monitored so an issue would be better

OlehKSS commented:
It is possible to create interactive plots with matplotlib itself, by default they have several control buttons unnecessary for a colorbar. I will check  how to customize it and whether it is possible at all. I have added an [issue](https://github.com/bloomberg/bqplot/issues/670) in bqplot repository similar to my question on gitter.

OlehKSS commented:
I have added a color bar with `bqplot`, using custom library build from the repository master branch, since the changes that I need have not been released yet. I have pushed my code to a [separate branch](https://github.com/OlehKSS/mne-gsoc2018-3d/tree/olehkss/27).

*Color bar*
![image](https://user-images.githubusercontent.com/16047980/42411190-6b3cdc64-8200-11e8-8062-709c8d45e73e.png)

OlehKSS commented:
Should I add the code for color bar creation here or should we merge this pull request and continue to work on color bar addition in a different pull request?

larsoner commented:
Adding it here is okay with me. But is the colorbar interactive, i.e., can you set the fmin/fmid/fmax for both one-sided (`hot`) and two-sided (`mne`) colormaps?

agramfort commented on mne_g3d/viz.py:
```diff
@@ -233,3 +236,319 @@ def plot_hemisphere_mesh(vertices,
+                    grade=None,
+                    smooth=smoothing_steps,
+                    subjects_dir=subjects_dir,
+                    subject_from=subject)
```
does this create an stc on the full resolution surface? this can create huge memory allocations. The upsampling to full resolution surface should be done on the fly as much as possible.

OlehKSS replied:
as far as I understand from @larsoner comments, this operation should perform downsampling using provided parameter `smoothing_steps`.

larsoner replied:
Reread my comment above -- I suggest to do something different, namely compute the sparse upsampling matrix (which is small in terms of memory) and do the upsampling dot operation on the fly for each time point as needed. What you have now is equivalent to doing the dot with all time points (which can be huge).

OlehKSS replied:
It will work fine for a given time instance, but with enabled time viewer it will be a problem. In that case `ipyvolume` expects me to send it all data at once, for every time instance while building a mesh.  Afterwards `ipyvolume` will select the appropriate data to show.

larsoner replied:
For many STCs this will not work. Upsampling needs to be done on the fly

OlehKSS replied:
I will check with `ipyvolume` folks, but so far I suppose it will require  some changes to `ipyvolume`.

OlehKSS replied:
I suppose we will need to have some callback function for time viewer

OlehKSS replied:
According to [this issue](https://github.com/maartenbreddels/ipyvolume/issues/92), this problem can be solved by creating custom animation tools using ipywidgets instead of using default ones. A new mesh for each time instance should be created and the old one closed.

larsoner replied:
It would be better just to update some data attribute rather than recreate the entire mesh

OlehKSS commented:
@larsoner the color bar is not interactive, for now it displays standard settings. It works in this way for `mne` and `hot` color maps. I am not sure how interactivity should be implemented like, do you have any suggestions? I have added the code for the color bar creation to this pull request.

agramfort commented:
@OlehKSS I discussed briefly with @larsoner and we agree that this code should end up in a pysurfer-like package and not directly in MNE. It will help spreading the tool at there is a lot more pysurfer users than MNE users.

We would shoot for a package called "ipysurfer" that mimics the API of PySurfer.

Then we'll integrate it into MNE.

what do you think?

OlehKSS commented:
Yeah, sure. In order to implement `ipysurfer`, should we create a different repository or continue to work in this one?

OlehKSS commented:
I have added an issue for creating ipysurfer package, see #29.

larsoner commented:
For the colorbar you should be able to use widgets to set the min/mid/max float values, and the colorbar and brain plot should both update

larsoner commented:
IIRC in some topomap viz we actually make it so you can click and drag in the colorbar to change limits but we can always add this later (and we want the float slider/entry boxes either way)

larsoner commented:
... you can get an idea for this already based on how PySurfer does it (like with most things). In this case, time_viewer=True shows float sliders.

OlehKSS commented:
I have changed upsampling, so now it uses sparse matrix and does upsampling for a selected time instance. Animation controls provided by `ipyvolume` also work, but now it is not that quick. I have added sliders for  changing control points of the color map. I could not find a `PySurfer` example of sliders for color map control, could you send a link to one? Here is a screenshot of what I have now:
![image](https://user-images.githubusercontent.com/16047980/42642084-b4e1bf22-85fe-11e8-8c55-8903c539f2db.png)

Do you have any naming suggestions for the sliders instead of `fmin/fmid/fmax`?
About changing API to PySurfer-like, do you have anything like UML table or something similar with information about high-level description of `PySurfer` project, so it would take less time for me to understand everything that should be done?

agramfort commented:
fmin/fmid/fmax is MNE standard naming convention.

regarding pysurfer I would try to replicate the examples:

https://pysurfer.github.io/auto_examples/index.html

especially

https://pysurfer.github.io/auto_examples/plot_basics.html#sphx-glr-auto-examples-plot-basics-py
https://pysurfer.github.io/auto_examples/plot_meg_inverse_solution.html#sphx-glr-auto-examples-plot-meg-inverse-solution-py
https://pysurfer.github.io/auto_examples/plot_fmri_activation.html#sphx-glr-auto-examples-plot-fmri-activation-py

OlehKSS commented:
Ok, so I will keep naming for sliders like it is now.
Is there any way to control color map parameters with `PySurfer`? I was looking for something similar to the sliders I've done for `fmin/fmid/fmax`.

larsoner commented:
See [my comment from a few days ago](https://github.com/mne-tools/mne-gsoc2018-3d/pull/25#issuecomment-403453249), use `time_viewer=True` it will show:

![screenshot from 2018-07-12 14-07-23](https://user-images.githubusercontent.com/2365790/42651291-fe62264a-85dc-11e8-9cb8-67be36a80e36.png)

larsoner commented:
(turns out they are not sliders but rather text entry boxes)

larsoner commented:
and as far as API is concerned, the API page for `Brain` should hopefully tell you what you need to know:

http://pysurfer.github.io/generated/surfer.Brain.html#surfer.Brain

OlehKSS commented:
I have changed sliders to input boxes and changed color map calculation slightly, so it will be more consistent with PySurfer-based plots.
![image](https://user-images.githubusercontent.com/16047980/42696417-bd5a89d2-86c0-11e8-93df-f6e5e0b93510.png)

agramfort commented:
looks good !

larsoner commented:
Does it work for `mne` / the two-sided colormap, too?

larsoner commented:
(If not, properly supporting two-sided data would be the next step)

larsoner commented:
FYI PySurfer does not support two-sided data so nicely, so what you create could in principle do a better job

OlehKSS commented:
@larsoner  I haven't tried two-sided data so far, could you share any examples of it? I guess it should work with some small changes , since I use double-sided material for the overlay, look [here](https://github.com/mne-tools/mne-gsoc2018-3d/pull/25/files#diff-ab01fb5e3a26fa4335027c877304018bL231).

larsoner commented:
Running this with `pick_ori='normal'` in the `apply_inverse` call will get you signed data:

https://github.com/mne-tools/mne-python/blob/master/tutorials/plot_mne_dspm_source_localization.py

OlehKSS commented:
Thanks, I obtained an inverse solution. I have some weird problems with `ipyvolume` at the moment, Jupyter notebooks could not find `ipyvolume.js`.  I will send you a plot when I find what's wrong with the packages.

OlehKSS commented:
I haven't found a solution to `ipyvolume` problems so far, but I have managed to run my code in the older virtual environment. Here is how signed data visualization looks like:
![image](https://user-images.githubusercontent.com/16047980/42817344-80b55cf0-89d6-11e8-955a-7561a3046b41.png)

As far as I understand, you are using both terms "signed data" and "two-sided data" interchangeably, right? Because for the first time I thought you were talking about double-sided mesh material...

larsoner commented:
Yes interchangably.

That colormap should have middle gray (0.5, 0.5, 0.5) in the middle, not black, can you fix it? Here is where PySurfer manually mixes:

https://github.com/nipy/PySurfer/blob/master/surfer/viz.py#L1943

OlehKSS commented:
Thanks, I could not figure out how you have it implemented before. Here is an exampled of fixed color map with grey in the middle:
![image](https://user-images.githubusercontent.com/16047980/42819767-2b5858a0-89dd-11e8-908d-13f1376d0507.png)

larsoner commented on mne_g3d/viz.py:
```diff
@@ -233,3 +237,402 @@ def plot_hemisphere_mesh(vertices,
+            bg_color = 0.5 * np.ones(3)
+            colors = np.array([255 * c[:-1] if c[-1] == 1 else
+                               255 * (c[:-1] * c[-1] + bg_color *
+                               (1 - c[-1])) for c in colors])
```
Iteration like this should be avoided if possible for performance (and readability) reasons, can you vectorize this using `ndarray` and broadcasting?

larsoner replied:
(In this case it's probably faster and simpler just to do all subtractions/multiplies rather than checking each alpha value, even if some of those alphas are 1)

OlehKSS replied:
yep, I guess so

larsoner commented:
Can you update the top-level comment with progress thus far? Do you think we should merge this now, or keep iterating on other things to add?

larsoner commented:
Colormap setting works, but when I click "Update mesh" it updates to show a different time point, too. Perhaps it's updating to show the `initial_time` point with the new colormap limits, rather than the currently selected time point?

Also in "movie mode" it is a little bit slow. Could you add to the list of TODO items (or just do it here) to have a [time_dilation](http://pysurfer.github.io/generated/surfer.Brain.html#surfer.Brain.save_movie) parameter that sets the ratio between time in the file and playback time, e.g. a factor of 4 (the default?) makes 1 sec of data play through in 10 seconds? This will require some math regarding `stc.tstep` to get it to happen correctly but will be nice for the user.

larsoner commented:
Also the colorbar itself does not update its `xlim` so it looks bad for some changes:

![screenshot from 2018-07-17 09-37-42](https://user-images.githubusercontent.com/2365790/42820833-1ea962da-89a5-11e8-9d64-e6ecaedddd37.png)

larsoner commented:
(and actually this looks incorrect, in addition to not updating the xlim)

larsoner commented:
Also the `colorbar` did not appear correct for me in two-sided mode (do I need `bqplot` master or something? I just `pip` installed), and updating the mesh breaks things -- you probably need to add a conditional in the update function for whether it's a one-sided or two-sided colormap:

![screenshot from 2018-07-17 09-40-54](https://user-images.githubusercontent.com/2365790/42821051-a8e262bc-89a5-11e8-9dbe-89e6588a86fe.png)

OlehKSS commented:
@larsoner Yes, you need to use `bqplot` from master branch, since they have not released yet what I need to build a color map

OlehKSS commented:
I have added an [issue for time dilation parameter](https://github.com/mne-tools/mne-gsoc2018-3d/issues/30). The big problem that I see now is delay in calculations. Previously all calculations were done once and then the data were sent to browser. Now, each time browser needs a new frame/plot to show it will call server for it. It significantly sowers animations and movies.

larsoner commented:
If you really want to make it fast, you can figure out how to get the sparse matrix into JavaScript and do the upsampling there, then use those values to update the mesh. Otherwise I don't see a way to avoid doing the calculation in Python.

I'm a bit surprised it's slow, though -- if I run this:
```
import time
import numpy as np
import mne

vertices = [np.arange(len(mne.read_surface(mne.utils.get_subjects_dir() +
                                           '/fsaverage/surf/%s.white'
                                           % hemi)[0]))
            for hemi in ('lh', 'rh')]
mm = mne.compute_morph_matrix('fsaverage', 'fsaverage',
                              [np.arange(10242), np.arange(10242)],
                              vertices, verbose=True)
data = np.random.randn(20484, 1)
t0 = time.time()
for _ in range(60):
    mm.dot(data)
print(time.time() - t0)
```
I get 30 ms. In other words, I can generate 60 "frames" of data (enough for 1 sec of smooth playback) in 30 ms. So the bottleneck is not in the upsampling. Where is it?

larsoner commented:
... although if I up the smoothing steps to 10 rather that `None` it takes 290 ms, which is getting a bit high. But still, something else must be taking up a lot more time if it's not smooth.

OlehKSS commented:
The communication is done via server, I would expect it to be the major bottleneck

larsoner commented:
Ahh I thought at one point that Maarten mentioned the serialization/transfer being fast, but maybe the server communication adds overhead, or I otherwise misunderstood him

OlehKSS commented:
I think server communication can't be fast enough to do a movie or etc. I have fixed bugs with initial_time selection and problems with signed data display after mesh update. I have also changed the way it calculates colors for the color bar; it doesn't use an if-statement anymore.

OlehKSS commented:
I have updated description, I did not add these parameters so far:
* foreground
* spacing
* title
* figure 

Should I add them and raise `NotImplementedError` or should I just omit them, as it is now?

larsoner commented:
Don't bother with `spacing`, we don't need it. For `figure` for now `NotImplementedError` is fine, unless you want to make it possible to take an `IPyvolume` figure and add the brain to it.

Also things like `views` as a list and `hemi='split'` are not supported, right? Those are important (probably next most important targets) so should be added, too. If you don't know what these do, try in PySurfer.

For `title` can you look into adding text annotations to IPyvolume figures? Is there no way to do it? Have you seen what it does in PySurfer so you know what I'm talking about? If you get this work then `foreground` is pretty much just for setting this text color right now.

OlehKSS commented:
These is no way for specifying the title of a plot using ipyvolume, I can do it by adding an html widget with text above the figure. Do you think we need `foreground` color regarding the fact that we do not any text on a figure in contrast with PySurfer?

About `hemi=split`, should those two separate meshes have the same color bar with the same settings or should they have different and separate from each other color  bars? 

For both `hemi=split` and `views=[...]` I will need to have several plots of the same mesh. The only way it can be done now is by creating several separate figure widgets. What do you think about this approach?

agramfort commented:
>
> These is no way for specifying the title of a plot using ipyvolume, I can
> do it by adding an html widget with text above the figure. Do you think we
> need foreground color regarding the fact that we do not any text on a
> figure in contrast with PySurfer?
>
html text is fine

> About hemi=split, should those two separate meshes have the same color
> bar with the same settings or should they have different and separate from
> each other color bars?
>
same

> For both hemi=split and views=[...] I will need to have several plots of
> the same mesh. The only way it can be done now is by creating several
> separate figure widgets. What do you think about this approach?
>
yes separate widgets with shared / linked variable for sync (eg the
colorbar changes and the time selection)

larsoner commented:
> These is no way for specifying the title of a plot using ipyvolume

There is no way to add text within an IPyvolume plot? That is all that is needed for a PySurfer-style title. If not, yes HTML is workable.

OlehKSS commented:
I have added `title` and `foreground parameters`, `split` and `views` as a list:
![image](https://user-images.githubusercontent.com/16047980/42992096-4387d100-8c10-11e8-83b9-688d8b1f148f.png)

The problem is that camera rotation is broken `ipyvolume` master, bug [#156](https://github.com/maartenbreddels/ipyvolume/issues/156), due to it `views` parameter doesn't work as expected. 
![image](https://user-images.githubusercontent.com/16047980/42992579-9a99eacc-8c11-11e8-89c7-f2b5d9d2fb4c.png)

agramfort commented:
can the title be added as a fig title next to the buttons?

OlehKSS commented:
`spacing` parameter was omitted and `figure` parameter was added, it will raise `NotImplementedError`

OlehKSS commented:
So, you want me to put title above the play-pause buttons, but below the figures. Am I correct?

OlehKSS commented:
![image](https://user-images.githubusercontent.com/16047980/42993334-c99c2d7e-8c13-11e8-8133-164e5debfda3.png)

agramfort commented:
I mean next to the ipyvolume blue buttons

OlehKSS commented:
No, for that it will require changes to ipyvolume, since those buttons are a part of the figure widget

OlehKSS commented:
@larsoner @choldgraf Do you have any additional comments to this pull request? Should we merge it?

larsoner commented:
For some reason in the "Plot source estimates with time viewer and color bar" cell I am unable to change `fmin` to any other value.

Sometimes updating fmin/fmid/fmax and clicking "update mesh" will cause the overlay to disappear completely, at least in the "signed data" section.

Do you experience either of these things?

OlehKSS commented:
For `fmin` , `fmid`and `fmax` I use minimum and maximum values returned by `limits_to_control_points` function in order to limit changes of these values. Do you think these values should not be limited and can be arbitrary? I have not experienced anything like you described. Do you have any errors in the console?

larsoner commented:
fmin/fmid/fmax should not be limited except being strictly increasing (or at least nondecreasing), and, for the `pos_lims` / MNE colormap case, all non-negative.

No errors in the console for the removals. I'm going to assume it's a weird problem with my system.

larsoner commented:
Great, let's keep going!
